BREAKING CHANGE: Refactor inputoptions component

blackfalcon opened the pull request:
# Alaska Airlines Pull Request

Update currently released ods-inputoptions with two new components to reduce the complexity of support. 

**Fixes:** #4, #9, #10

## Summary:

The core features of ods-inputoptions will be preserved and with this update will release two new components for use. ods-inputoption-group and ods-inputoption.

ods-inputoption-group will be a wrapper component used within the HTML and ods-inputoption cab be used repeatedly as elements within the slot of ods-inputoption-group. This will allow for a less complex API of the component as well as give developers direct access to the inner components. 

Worked within the scope of this update also includes support for a light DOM version of the component. 

## Type of change:

Please delete options that are not relevant.

- [x] New capability 
- [x] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate) 


## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._ 

**Thank you for your submission!**<br>
-- Orion Design System Team

c-n-d committed:
refactor: split input options and group into separate components

c-n-d committed:
feat: light dom input option

c-n-d committed:
refactor: add notes to light dom element; ifdefined guard

c-n-d committed:
fix: update .stylelintrc to allow ods-inputoptions-group custom element

c-n-d committed:
test: add @open-wc/testing; create tests for inputoption/light dom/group

c-n-d committed:
feat: splits ods-inputoptions in to multiple components

1. Update demo page to use new components
1. Modify tests to cover addition behavioral cases
1. Change `id` property to `inputId`
1. Remove capComponentLoop functionality to limit number of elements in a group

TODOs:
1. Styling for horizontal inputs (due to inputs now being in a slot)
1. Remove ods-inputoptions element, related tests, and demo components.
1. Update documentation for new componenets

BREAKING CHANGE

blackfalcon commented on src/ods-inputoption-group.js:
```diff
@@ -0,0 +1,82 @@
+
+      <div
+        @input="${this._handleInput}"
+        class="${classMap(groupClasses)}">
```
I really like this. Just started looking at classMap too and seeing how other components can be refactored.

c-n-d committed:
feat: add checks for defined labels and input options

TODOs:

1. The 'supports tab and arrow group traversal' in ods-inputoption-group.test.js is not currently successful in testing tab traversal through the input group

Commit:
fix: add tab functionality for accessibility

Commit:
fix: update todo comments to reflect reality

AuroDesignSystem reviewed (changes requested):
Overall, looking great. But there is an issue with the a11y tabbing states. The user will tab INTO a radio button set and use the ARROW KEYS to move between buttons. 

Then the user will tab OUT of the set. 

The way this is currently working, the a11y user is trapped.  Please review the a11y interaction on the MASTER branch. 

There is the same issue with checkboxes. 

Tabbing appears to be totally broken with light DOM versions?

AuroDesignSystem commented on demo/index.html:
```diff
@@ -13,6 +13,9 @@
       import '@polymer/iron-demo-helpers/demo-snippet';
     </script>
     <script type="module" src="../src/ods-inputoptions.js"></script>
```
ods-inputoptions should be removed from this version. This is a BREAKING CHANGE and there is no need to support this version any longer.

AuroDesignSystem reviewed (changes requested):
There seems to be a general issue with the tabbing. In the examples on the demo page, when I click on the LABEL, the first option is not selected? 

Please see https://auro-prod-site-w2.azurewebsites.net/ods-options for the initial version of this component. 

There seems to be some wonkyness with the CSS selector.

blackfalcon committed:
fix: issues with style dependencies

Commit:
fix: remove tab hijacking for radio buttons

Commit:
feat: add ods-inputoption-radio-group component for radio groups

Commit:
feat: add ods-inputoption-checkbox-group component for checkbox groups

Commit:
style: address light DOM UI

blackfalcon committed:
fix: update a11y UX

BREAKING CHANGE: this PR re-writes the component API requireing a MAJOR
release.

ecommerceUXdesign commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@alaskaairux/ods-inputoptions/v/2.0.0)
- [GitHub release](https://github.com/AlaskaAirlines/ods-inputoptions/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
